Add change stream support for projections

sandrokeil opened the pull request:
Patch file for duplicate check test.

[Test_duplicate_change_stream_detection_patch.txt](https://github.com/sandrokeil/mongodb-event-store/files/2384059/Test_duplicate_change_stream_detection_patch.txt)

[Test_duplicate_change_stream_detection_wrong_detection_patch.txt](https://github.com/sandrokeil/mongodb-event-store/files/2394054/Test_duplicate_change_stream_detection_wrong_detection_patch.txt)

sandrokeil committed:
Add change stream support for projections

- https://docs.mongodb.com/manual/changeStreams/

sandrokeil commented on src/Projection/ProcessEvents.php:
```diff
@@ -0,0 +1,208 @@
+                    // event already processed
+                    if ($eventTimestamp <= $streamTimestampsStart[$streamName]
+                        && $event['fullDocument']['_id'] <= $this->streamPositions[$streamName]
+                    ) {
```
Already processed event is hard to test and I'm not sure if detects all cases.

codeliner replied:
maybe it helps to draw some scenarios using draw.io?
The implementation looks correct

sandrokeil replied:
@codeliner I've added another patch (Test_duplicate_change_stream_detection_wrong_detection_patch.txt). It looks like this can not be fixed. You should see three duplicate detections, which is wrong. Maybe you have to change the `usleep` setting or run it more than once.

codeliner commented on src/Projection/ProcessEvents.php:
```diff
@@ -0,0 +1,208 @@
+                $collectionNames[$this->persistenceStrategy->generateCollectionName(new StreamName($streamName))] = $streamName;
+            }
+            // initialize stream to get new events during processing current events
+            $changeStream = $this->client->selectDatabase($this->database)->watch([
```
What happens In a replay scenario that takes one hour or so? This could cause memory issues, or?

sandrokeil replied:
Good point, we can add a [`batchSize`](https://docs.mongodb.com/manual/reference/method/db.collection.watch/) to avoid it.

codeliner reviewed (changes requested):
Awesome work

sandrokeil committed:
Add gap history

sandrokeil commented on tests/Projection/MongoDbEventStoreReadModelProjectorSimpleStreamStrategyTest.php:
```diff
@@ -0,0 +1,94 @@
+        $this->assertSame(22, $result['state']['aggregate_positions'][19]);
+        // events from above are saved/processed as last but have first position
+        $this->assertSame(1, $result['state']['aggregate_positions'][20]);
+        $this->assertSame(2, $result['state']['aggregate_positions'][21]);
```
@codeliner This could be a problem if aggregate root does not check version increment. It's not a version collision. But it depends on events of the same aggregate (race condition).

sandrokeil commented:
You can also play with the Event Store Benchmark 

https://github.com/sandrokeil/event-store-benchmarks/tree/feature/mongodb

sandrokeil committed:
Update to Event Store 7.4 (#2)

sandrokeil committed:
Retry MongoDB commit and ignore errors on rollback

- http://php.net/manual/de/mongodb-driver-session.committransaction.php
- http://php.net/manual/de/mongodb-driver-session.aborttransaction.php
- https://docs.mongodb.com/manual/core/transactions/#retry-transaction
